pkg/kubelet/kubelet_pods.go

```diff
@@ -343,7 +343,14 @@ func ensureHostsFile(fileName string, hostIPs []string, hostName, hostDomainName
 		// if Pod is not using host network, create a managed hosts file with Pod IP and other information.
 		hostsFileContent = managedHostsFileContent(hostIPs, hostName, hostDomainName, hostAliases)
 	}
-
+	readContent, err := ioutil.ReadFile(fileName)
+	if err == nil {
+		oldCheckSum := sha512.Sum512_256(readContent)
+		newCheckSum := sha512.Sum512_256(hostsFileContent)
+		if oldCheckSum == newCheckSum {
+			return nil
+		}
+	}
 	return ioutil.WriteFile(fileName, hostsFileContent, 0644)
```

Comment:
Let us also ensure that the write is atomic.

Reply:
the `ensureHostsFile` is called inside the container `start ` func, each container start is called serialized in a pod, so it should safe enough and there is no concurency for multi goroutines for a single pod

Reply:
Although each container start is called serialized in a pod, but since multiple containers share the same /var/lib/kubelet/pods/xxxxx/etc-hosts file, if the container that has been started is using the /etc /hosts file, however a new container is starting and generating a new etc-hosts file, this will affect the operation that is using the /etc/hosts file.
Such as using hostname -f to resolve FQDN, as #111849 describe.